### PR TITLE
Add weekly & yearly summaries

### PR DESCRIPTION
**Pull request by GitOldGrumpy**
Implements the two missing parts of summaries. 
- There are two new lint errors that I don't understand. If someone could explain what I've missed.
- If anyone has a car older than a year could they check if they can receive data from over 12 months ago.

Finally, anyone think we need a helper function, where you can ask for THIS_WEEK, LAST_WEEK, THIS_MONTH, LAST_MONTH, THIS_YEAR, LAST_YEAR or is it just me that uses this functionality?

### CODE REVIEW

File: mytoyota/models/endpoints/trips.py
Comment (CM000n) [inline mytoyota/models/endpoints/trips.py:48]: `mytoyota/models/endpoints/trips.py:48:12: E1101: Instance of '_SummaryBaseModel' has no 'length' member (no-member)`

Mhmm, really strange , since `length` is clearly a member of `_SummaryBaseModel`.

**Comment by CM000n** [comment]
> ## [Codecov](https://app.codecov.io/gh/DurgNomis-drol/mytoyota/pull/265?src=pr&el=h1&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=Simon+Hansen) Report
> Attention: `71 lines` in your changes are missing coverage. Please review.
> 
> > Comparison is base [(`057595f`)](https://app.codecov.io/gh/DurgNomis-drol/mytoyota/commit/057595fa5d568bcce8afa0643f36fb918a918623?el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=Simon+Hansen) 75.23% compared to head [(`d62ee54`)](https://app.codecov.io/gh/DurgNomis-drol/mytoyota/pull/265?src=pr&el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=Simon+Hansen) 71.65%.
> > Report is 3 commits behind head on master.
> 
> [Files](https://app.codecov.io/gh/DurgNomis-drol/mytoyota/pull/265?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=Simon+Hansen)	Patch %	Lines
> [mytoyota/models/vehicle.py](https://app.codecov.io/gh/DurgNomis-drol/mytoyota/pull/265?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=Simon+Hansen#diff-bXl0b3lvdGEvbW9kZWxzL3ZlaGljbGUucHk=)	4.44%	[43 Missing ⚠️ ](https://app.codecov.io/gh/DurgNomis-drol/mytoyota/pull/265?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=Simon+Hansen)
> [mytoyota/models/endpoints/trips.py](https://app.codecov.io/gh/DurgNomis-drol/mytoyota/pull/265?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=Simon+Hansen#diff-bXl0b3lvdGEvbW9kZWxzL2VuZHBvaW50cy90cmlwcy5weQ==)	12.50%	[28 Missing ⚠️ ](https://app.codecov.io/gh/DurgNomis-drol/mytoyota/pull/265?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=Simon+Hansen)
> Additional details and impacted files
> [☔ View full report in Codecov by Sentry](https://app.codecov.io/gh/DurgNomis-drol/mytoyota/pull/265?src=pr&el=continue&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=Simon+Hansen). 📢 Have feedback on the report? [Share it here](https://about.codecov.io/codecov-pr-comment-feedback/?utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=Simon+Hansen).

I think you should include some additional unit tests @GitOldGrumpy , if possible

**Comment by CM000n** [comment]
> Finally, anyone think we need a helper function, where you can ask for THIS_WEEK, LAST_WEEK, THIS_MONTH, LAST_MONTH, THIS_YEAR, LAST_YEAR or is it just me that uses this functionality?

I also think that this could be extremely helpful. 👍🏻 Especially in view of the statistics sensors for the Home Assistant integration.

File: mytoyota/models/vehicle.py
Comment (CM000n) [inline mytoyota/models/vehicle.py:251]: The whole `get_summary` code is complex due to repetitive sorting operations, redundant initializations of the return list ret, and unnecessary checks for None in accumulation logic.

I think it can be imporved by removing the redundant initializations of ret outside of the conditional branches since it's re-initialized in each branch.
You can also consolidate the sorting of resp.payload.summary before the conditional branches to avoid sorting in each branch. And simplify the accumulation logic by initializing build_hdc and build_summary once at the start of the loop instead of checking for None every time.
The functions for collecting the statistical data within the `for summary...` loops should also be outsourced to independent functions.

**Review by CM000n — CHANGES_REQUESTED** [review]

File: mytoyota/models/endpoints/trips.py
Comment (CM000n) [inline mytoyota/models/endpoints/trips.py:39]: Can you please explain a little why this is needed?

File: mytoyota/models/vehicle.py
Comment (CM000n) [reply mytoyota/models/vehicle.py:343]: I've switched to ruff for linking, which I've wanted to do for a long time anyway. As a result, this comment has become obsolete.

File: mytoyota/models/endpoints/trips.py
Comment (CM000n) [reply mytoyota/models/endpoints/trips.py:51]: I've switched to ruff for linking, which I've wanted to do for a long time anyway. As a result, this comment has become obsolete.

File: mytoyota/models/endpoints/trips.py
Comment (CM000n) [inline mytoyota/models/endpoints/trips.py:13]: Could you please explain what exactly this is needed for.
Please also add a few unit tests to maintain the coverage. I think these could also contribute to the explanation.

File: mytoyota/models/endpoints/trips.py
Comment (CM000n) [inline mytoyota/models/endpoints/trips.py:108]: Could you please explain what exactly this is needed for.
Please also add a few unit tests to maintain the coverage. I think these could also contribute to the explanation.

File: mytoyota/models/endpoints/trips.py
Comment (CM000n) [inline mytoyota/models/endpoints/trips.py:41]: Could you please explain what exactly this is needed for.
Please also add a few unit tests to maintain the coverage. I think these could also contribute to the explanation.

**Review by CM000n — CHANGES_REQUESTED** [review]

File: mytoyota/models/endpoints/trips.py
Comment (GitOldGrumpy) [reply mytoyota/models/endpoints/trips.py:13]: When creating the `__add__` functions I found myself writing this set of code over and over to deal with the situation where either this and/or that are None. If either side is None(but not both) you get an exception. In these situations we just want the result to be the side that is not None.

File: mytoyota/models/endpoints/trips.py
Comment (GitOldGrumpy) [reply mytoyota/models/endpoints/trips.py:108]: These are required as the only way to get the Monthly and Yearly summaries is too add together either the Daily summaries or the Monthly Summaries, seemed the best place to add them.

File: mytoyota/models/vehicle.py
Comment (GitOldGrumpy) [reply mytoyota/models/vehicle.py:230]: The logic is complicated. I have attempted to improve it(in changes shortly) and have added more comments.

build_hdc and build_summary can not be initialised once. For example, if you are building weekly summaries for multiple months you dont know where the week ends, it could be in the summary loop or the histogram loop. To deal with this I reset them to None each time we find a Sunday so we can restart building the next week.

**Comment by CM000n** [comment]
> * If anyone has a car older than a year could they check if they can receive data from over 12 months ago.

Hi @GitOldGrumpy, my car is from 2019. It seems that I can only retrieve data that goes back a maximum of 364 days. For today's date (30.12.2023), for example, the oldest trip for which I receive information is 31.12.2022.

File: mytoyota/models/endpoints/trips.py
Comment (CM000n) [reply mytoyota/models/endpoints/trips.py:13]: As far as I know, there is no built-in python functionality for this. I would have solved it slightly differently, but this is more or less similar to your approach:

```python
def add_with_none(a: Any, b: Any) -> Any:
    return a + b if a is not None and b is not None else a or b
```
Since the function can be used for anything, I still think that it is better placed somewhere in the `utils` than in the trips endpoints models.

I generally think that some more specific unit tests can already be helpful in the work before we move on to writing larger integration tests. :blush:

File: mytoyota/models/vehicle.py
Comment (CM000n) [inline mytoyota/models/vehicle.py:230]: Please do not simply use `# noqa: PLR0912, PLR0915` here. Instead, we should try to improve the code quality as described.

**Review by CM000n — CHANGES_REQUESTED** [review]

File: mytoyota/models/vehicle.py
Comment (CM000n) [inline mytoyota/models/vehicle.py:305]: Why is this called `today`? 
This value is not actually today, but can change with each loop over a histogram and only shows the date present in the respective histogram, right?

File: mytoyota/models/vehicle.py
Comment (CM000n) [inline mytoyota/models/vehicle.py:356]: Hi @GitOldGrumpy, my car is from 2019. It seems that I can only retrieve data that goes back a maximum of 364 days. For today's date (30.12.2023), for example, the oldest trip for which I receive information is 31.12.2022.

File: mytoyota/models/vehicle.py
Comment (CM000n) [reply mytoyota/models/vehicle.py]: I'm just thinking out loud and need to give it some more thought and try it myself: But wouldn't it be possible to make the code more consistent and uniform if we built some helper functions for week, month and year?
These helper functions are given a start date and end date, which they use to generate the weeks, months and years within them with their respective start and end dates.
With these generated start and end dates of the respective time periods, it should then be possible to filter the summary data consistently, right?

File: mytoyota/models/vehicle.py
Comment (GitOldGrumpy) [reply mytoyota/models/vehicle.py:356]: I expected as much but dont think it makes anything different to the code.

File: mytoyota/models/vehicle.py
Comment (CM000n) [reply mytoyota/models/vehicle.py]: Something like this for example for weeks:

```python
from datetime import datetime, timedelta

def kalenderwochen_dict(startdatum, enddatum):
    start_date = datetime.strptime(startdatum, "%Y-%m-%d")
    end_date = datetime.strptime(enddatum, "%Y-%m-%d")

    kalenderwochen_dict = {}

    current_date = start_date
    while current_date <= end_date:
        kw_start = current_date - timedelta(days=current_date.weekday())
        kw_end = kw_start + timedelta(days=6)

        kalenderwochen_dict[f"{current_date.year}-{current_date.isocalendar()[1]}"] = {
            'startdatum': kw_start.strftime("%Y-%m-%d"),
            'enddatum': kw_end.strftime("%Y-%m-%d")
        }

        current_date += timedelta(weeks=1)

    return kalenderwochen_dict

# Beispielaufruf:
startdatum = "2023-01-01"
enddatum = "2023-12-31"
resultat = kalenderwochen_dict(startdatum, enddatum)
print(resultat)
```

File: mytoyota/models/vehicle.py
Comment (GitOldGrumpy) [reply mytoyota/models/vehicle.py:305]: Yes its "today" in relation to the loop. I've updated the names.

File: mytoyota/models/vehicle.py
Comment (GitOldGrumpy) [reply mytoyota/models/vehicle.py]: I did think about this, but didn't implement as you use different parts of the summary response depending on what has been requested.
- Daily: Simply pull out the data returned in the histograms, no maths required.
- Weekly: Add together the days that make up a week from the histograms.
- Monthly: Simply pull out the monthly summary supplied in the data, no maths required. NOTE: No need to look at the histograms.
- Yearly: Add together the monthly summaries that make up the year. NOTE: No need to look at the histograms.

Yes, it could all be done with the histograms but having been provided with the Monthly data I made use of it as opposed to doing a lot of extra calculations on the histogram data.

Comment (CM000n) [reply mytoyota/models/vehicle.py]: Good point! You're right, we shouldn't do manual calculations where they're not absolutely necessary.
Your recent changes have also made the code much better structured and easier to understand.

**Review by CM000n — CHANGES_REQUESTED** [review]

File: mytoyota/models/vehicle.py
Comment (CM000n) [inline mytoyota/models/vehicle.py:351]: I think all the summary functions should be given a more meaningful function name. At the moment, I can't tell directly from the function name what is being done here.
So instead of `_daily_summaries` it is better to use `_generate_daily_summaries` or similar.

**Review by CM000n — CHANGES_REQUESTED** [review]

File: mytoyota/models/vehicle.py
Comment (CM000n) [reply mytoyota/models/vehicle.py]: OK, after thinking about it again, the question has arisen as to what we want to return and what the user expects.
At the moment, we deliver all units for all full months of the start and end date that have already been pre-calculated by Toyota, especially for the monthly values, right?
As a user, however, I would expect that when a start and end date is specified, exactly this time span is used for an aggregation at monthly level, without the additional days before and after for the full months.

File: mytoyota/models/vehicle.py
Comment (GitOldGrumpy) [reply mytoyota/models/vehicle.py:351]: Not being a Python programmer I never think to use list comprehensions unless its a simple use case.

File: mytoyota/models/vehicle.py
Comment (GitOldGrumpy) [reply mytoyota/models/vehicle.py:371]: Does this give the week starting the Monday or the week starting on the day of the first histogram?

Comment (CM000n) [reply mytoyota/models/vehicle.py:371]: By default, `.span("week")` always gives you the week associated with a date, starting with Monday.
Screenshot 2024-01-02 134738

File: mytoyota/models/vehicle.py
Comment (CM000n) [inline mytoyota/models/vehicle.py:270]: I don't think this comment belongs here anymore?

**Review by CM000n — APPROVED** [review]

**Comment by CM000n** [comment]
LGTM now 🎉 
There is only one small issue left with what I think is a misplaced comment.
After that you can actually merge. 😊

**Comment by GitOldGrumpy** [comment]
Removing the comment has removed the merge approval. If you want to approve and then merge. Thanks.

**Review by CM000n — APPROVED** [review]

**Pull request merged**
